### ISSUE TITLE
User error in PodTemplate can effectively disable ECK

### ISSUE THREAD

**Issue by pebrc** [elastic/cloud-on-k8s#3021]
In https://github.com/elastic/cloud-on-k8s/issues/1822 we removed the validations on the `podTemplate` because:
* `containers` must not be empty but we want it to be empty in many cases when users want to override some other part of the template
* we also ran into issues with `kubectl apply` iirc because of the size of our manifests

But we are now in a situation where users can for example set `memory: ""` which the kube apiserver will accept due to the misssing validation but which will then make the operator get stuck in an endless loop of deserialization errors because this is not a valid quantity value. 

We should reevaluate whether we can re-introduce `podTemplate` validation in a customized form that caters to our needs (relaxed requirements on `containers` for example`)

**Comment by barkbay** [comment]
Isn't the webhook supposed to detect the problem and reject the manifest ? Or do we want something in addition to the webhook ?

The only case where it would not work is on Openshift 3.11 where the admission controller is disabled by default IIRC.

**Comment by pebrc** [comment]
> Isn't the webhook supposed to detect the problem and reject the manifest ? Or do we want something in addition to the webhook ?

Which validation would catch that?

**Comment by barkbay** [comment]
> Which validation would catch that?

None, the object must be deserialized before being validated by the webhook server. I would expect the deserialization to fail and thus the manifest to be rejected.

**Comment by barkbay** [comment]
I would expect this function to fail while decoding the request: https://github.com/kubernetes/kubernetes/blob/9eced040142454a20255ae323279a38dc6b2bc1a/staging/src/k8s.io/apimachinery/pkg/api/resource/quantity.go#L644

**Comment by pebrc** [comment]
You are right! But I guess we still want to do something about this as we cannot rely on the webhook being present in all cases. 

I wonder though that now that we are doing concurrent reconciles that the operator will still be able to make progress despite this error on other resources that are reconciled on the other go routines.

**Comment by anyasabo** [comment]
>I wonder though that now that we are doing concurrent reconciles that the operator will not be fully blocked by this error. The other goroutines should be able to reconcile unaffected resources

This doesn't appear to be the case in my test. None of the ES resources get reconciled as soon as there is one malformed resource. That error appears to be thrown during the ListAndWatch before it even starts dispatching to a reconcile loop.

I also noticed that it does not appear to use our logger (since it's not json formatted)

**Comment by anyasabo** [comment]
This seems important to solve but I'm not sure how to work around the max annotation length. I think the real fix is using server side apply, but:

- it is not the default in kubectl, but is available via flag
- it is beta in 1.16 but has rough edges
- it is in beta2 in 1.18
- not planned for GA in 1.19

Recent-ish status article [here](https://kubernetes.io/blog/2020/04/01/kubernetes-1.18-feature-server-side-apply-beta-2/) and enhancement issue https://github.com/kubernetes/enhancements/issues/555. Since we are still supporting 1.11 it seems like we have a while to go before we are able to fix it across the board. We may be able to go back to separate manifests

I tried it in 1.16 and the rough edge described in that article is indeed present. I tried applying our CRDs as is and it errored out:
```
10:36 $ k apply --server-side -f config/crds/all-crds.yaml 
Apply failed with 6 conflicts: conflicts with "before-first-apply" using apiextensions.k8s.io/v1beta1:
- .spec.group
- .spec.scope
- .spec.names.kind
- .spec.names.plural
- .status.conditions
- .status.storedVersions
Please review the fields above--they currently have other managers.
```

GKE only supports up to 1.16 now, would need to switch platforms to test 1.18 behavior.

**Comment by anyasabo** [comment]
1.18 is on the GKE rapid channel now so we could theoretically start to experiment with SSA
